### PR TITLE
Lazy submodules import

### PR DESCRIPTION
**Pull request by ericpre**
Since support for python 3.6 have been dropped, we can now use [PEP562](https://www.python.org/dev/peps/pep-0562/), which allows importing submodules lazily without much boilerplate code and without adding a dependency! @thomasaarholt, you may like this PR!
Alternative to #2636.

### Progress of the PR
- [x] Use lazy import for `signals`, `model`,
- [x] remove importing gui extension, since this is not necessary anymore, 
- [x] fix autocompletion of `hs.signals` by adding `__dir__`,
- [x] fix docstring components and api doc - see for example: [`hs.model`](https://hyperspy--2850.org.readthedocs.build/en/2850/api/hyperspy.utils.model.html), [`hs.model.components1D`](https://hyperspy--2850.org.readthedocs.build/en/2850/api/hyperspy.components1d.html#module-hyperspy.components1d) and [`hs.model.components1D.Exponential`](https://hyperspy--2850.org.readthedocs.build/en/2850/api/hyperspy._components.exponential.html#hyperspy._components.exponential.Exponential),
- [x] use pint UnitRegistry singleton,
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
From a jupyter notebook
```python
In [1]: %time import hyperspy.api as hs
CPU times: user 175 ms, sys: 22.3 ms, total: 198 ms
Wall time: 196 ms
```

```
python -X importtime -c "import hyperspy.api" 2> hs.txt && tuna hs.txt
```
has for output:
image

### CODE REVIEW

**Comment by thomasaarholt** [comment]
!!!
So much faster!

**Comment by ericpre** [comment]
> !!! So much faster!

Yes, indeed, I was also annoyed with this for a long time - as you would have been, I guess... and PEP562 allows to use a standardised approach and make it easy to implement and maintain.

@jlaehne, any chance, you can redo your benchmark in https://github.com/hyperspy/hyperspy/pull/2636#issuecomment-772546284 with this PR?

**Comment by jlaehne** [comment]
I reproduced the benchmark from my last comment in https://github.com/hyperspy/hyperspy/pull/2636#issuecomment-772546284

```
# RELEASE next minor (after jupyter restart)
CPU times: user 4.36 s, sys: 300 ms, total: 4.66 s
Wall time: 5.17 s
```

```
# pr2850 (after jupyter restart)
CPU times: user 407 ms, sys: 42.8 ms, total: 450 ms
Wall time: 470 ms
```

So more than factor 3 gain compared with #2636

File: hyperspy/_components/gaussianhf.py [hunk 1]
```diff
@@ -64,15 +65,21 @@ class GaussianHF(Expression):
...
+        Extra keyword arguments are passed to the
+        :py:class:`~._components.expression.Expression` component.
 
-
-    The helper properties `sigma` and `A` are also defined for compatibility
-    with `Gaussian` component.
+    Attributes
+    ----------
+    A : float
+        Convenience attribute to get and set the height and defined for
```
Comment (jlaehne) [inline hyperspy/_components/gaussianhf.py]: replace `and` by `,`

File: hyperspy/_components/gaussianhf.py [hunk 2]
```diff
@@ -64,15 +65,21 @@ class GaussianHF(Expression):
...
+        compatibility with `Gaussian` component.
+    sigma : float
+        Convenience attribute to get and set the width and defined for
```
Comment (jlaehne) [inline hyperspy/_components/gaussianhf.py]: replace `and` by `,`

File: hyperspy/utils/__init__.py [hunk 3]
```diff
@@ -85,3 +85,21 @@ def print_known_signal_types():
...
+    'plot',
+    'print_known_signal_types',
+    'roi',
+    'samfire',
+    'stack',
+    'transpose',
+    ]
+
+
+def __dir__():
```
Comment (jlaehne) [inline hyperspy/utils/__init__.py:98]: There are a number of such codecov warnings

Comment (ericpre) [reply hyperspy/utils/__init__.py:98]: Yes, I need to add tests as done in https://github.com/hyperspy/hyperspy_gui_ipywidgets/pull/40 and https://github.com/hyperspy/hyperspy_gui_traitsui/pull/41

**Comment by ericpre** [comment]
@jlaehne, would you be interested to make another review?

File: hyperspy/__init__.py [hunk 4]
```diff
@@ -41,7 +42,26 @@
...
+    }
+
+
+def __getattr__(name):
+    if name in __all__:
+        if name in _import_mapping.keys():
+            import_path = 'hyperspy' + _import_mapping.get(name)
+            return getattr(importlib.import_module(import_path), name)
+        else:
+            return importlib.import_module("." + name, 'hyperspy')
```
Comment (jlaehne) [inline hyperspy/__init__.py:69]: Still one codecov warning. I guess we should try to have `__init__.py` covered as good as possible.

Comment (jlaehne) [reply hyperspy/__init__.py:69]: A similar line in `api.py` (l. 47) seems to be covered.

**Review by jlaehne — APPROVED** [review]
Otherwise looks good from what I can see, but I wouldn't mind if a second pair of eyes had a look (@thomasaarholt ?).

File: hyperspy/__init__.py (hunk 4 above)
Comment (ericpre) [reply hyperspy/__init__.py:69]: I added a comment, explaining that we can't get covered with the test suite but if it breaks, it is noticeable.

**Comment by ericpre** [comment]
> Otherwise looks good from what I can see, but I wouldn't mind if a second pair of eyes had a look (@thomasaarholt ?).

@thomasaarholt, would you have the time to have a look at this PR, please?

**Comment by thomasaarholt** [comment]
Huge improvement in import time. On my M1 macbook I go from 2.54 sec to 0.1 sec. 

The code looks very good. My main concern was whether autocompletion would still work, and it does :)

This will hopefully make it even easier and fun to develop hyperspy code! 🚀

**Review by thomasaarholt — APPROVED** [review]

**Comment by ericpre** [comment]
Thanks @thomasaarholt for the review!

**Comment by ericpre** [comment]
I have merged it so that it can get widely used before the 1.7 release.

**Pull request merged**